Client Refactoring

mcorbin opened the pull request:
Hello,

I have done a a lot of refactoring on the Riemann Go Client:

- Deleted/simplified a lot of code
- Modified the Event => proto.Event conversion
- Separate UDP and TCP clients
- Add units/integration tests.
- Reorganize files
- ...

Only the tls client is missing, but i will work on it soon. Before that, i need reviews :)
I have one problem with one test failing sometime (tmarshal_test.go:67). I printed the protoRes and protoTest variables when the test fail. Here is the output :

```
time:100 service:"foobar" host:"baz" tags:"hello" ttl:10 attributes:<key:"bar" value:"baz" > attributes:<key:"foo" value:"bar" > metric_d:100.1

time:100 service:"foobar" host:"baz" tags:"hello" ttl:10 attributes:<key:"foo" value:"bar" > attributes:<key:"bar" value:"baz" > metric_d:100.1
```
I think the problem is the attributes order. I don't really know how to fix it. Otherwise, everything seems to work.

Can someone with Go experience review this PR ?

cc @jamtur01

bracki commented:
Regarding the tests I would remove the "Unit"/"Integ" from the test names and move the integration tests to files of their own. E.g. `foo_integration_test.go`. To these I'd add a build flag, so Integration tests can be run with `go test -t integration `. I'd also recommend automating the integration tests within the Makefile. Download and start Riemann, run all tests, tear down again.

bracki commented:
See http://peter.bourgon.org/go-in-production/#testing-and-validation for details on the build tag mentioned above.

mcorbin commented:
Thanks for your review @bracki . I modified the integration tests (using tags) and the Makefile.
I also fixed the marshal test (i now sort attributes before converting Event to protobuf).

bracki commented:
Cool. LGTM.

mcorbin commented:
@jamtur01 Ok for you ?

jamtur01 commented:
Looks good! Awesome work!

mcorbin commented:
@jamtur01 thanks ;) i will take a look at heapster now
